File: Concurrency/1242. Web Crawler Multithreaded.py
"""
The ProcessPoolExecutor class is an Executor subclass that uses a pool of processes to execute calls asynchronously.

 ProcessPoolExecutor uses the multiprocessing module, which allows it to side-step the Global Interpreter Lock 
 
 but also means that only picklable objects can be executed and returned.


An Executor subclass that executes calls asynchronously using a pool of at most max_workers processes. 
If max_workers is None or not given, it will default to the number of processors on the machine. 
If max_workers is less than or equal to 0, then a ValueError will be raised. 

bookkeeping of the seen urls are centralized in the main thread, therefore eliminates the need for a lock,
 which is necessary if worker threads also bookkeep seen urls.



Trying to understand this a bit especially inside for loop, here we are blocking the main thread till the url which is at the beginning of the deque 

is fully crawled. Won't there be cases where subsequent ones are done may be in which case you will be done faster crawling the ones

 from the ones with result as completed? Should we be using something like this?
 
  for task in concurrent.futures.as_completed(tasks):

"""
from concurrent import futures

class Solution:
    def crawl(self, startUrl: str, htmlParser: 'HtmlParser') -> List[str]:
        hostname = lambda url: url.split('/')[2]
        seen = {startUrl}
    
        with futures.ThreadPoolExecutor(max_workers=16) as executor:
            tasks = deque([executor.submit(htmlParser.getUrls, startUrl)])
            while tasks:
                for url in tasks.popleft().result():
                    if url not in seen and hostname(startUrl) == hostname(url):
                        seen.add(url)
                        tasks.append(executor.submit(htmlParser.getUrls, url))
        
        return list(seen)


https://leetcode.com/problems/web-crawler-multithreaded/solutions/929995/python3-using-concurrent-futures-threadpoolexecutor-with-error-handling/

from concurrent.futures import ThreadPoolExecutor, as_completed
class Solution:
    def process_url(self, url):
        urls = set(u for u in self.htmlParser.getUrls(url) if u.startswith(self.hostname))
        not_visited = urls - self.seen
        for u in not_visited:
            self.submit_to_executor(u)

    def submit_to_executor(self, url):
        self.seen.add(url)
        self.pending.append(self.executor.submit(self.process_url, url))

    def crawl(self, startUrl, htmlParser):
        self.seen = set()
        self.hostname = '/'.join(startUrl.split('/', 3)[:3])
        self.htmlParser = htmlParser
        self.pending = []

        with ThreadPoolExecutor(max_workers=64) as self.executor:
            self.submit_to_executor(startUrl)
            while self.pending:
                pending_so_far, self.pending = self.pending, []
                for fut in as_completed(pending_so_far):
                    if e := fut.exception():
                        raise RuntimeError("Future failed with an exception") from e

        return self.seen



import threading 
import collections


class Solution:
    def __init__(self):
        self.NUM_Threads = 4
        self.__cv = threading.Condition()
        self.__q = collections.deque()
        self.__working_count = 0
        self.__lookup = set()
        
    def crawl(self, startUrl: str, htmlParser: 'HtmlParser') -> List[str]:
        hostname = lambda url: url.split('/')[2]
        self.__q.append(startUrl)
        self.__lookup.add(startUrl)
        
        def worker(htmlParser):
            while True:
                with self.__cv:
                    while not self.__q and self.__working_count != 0:
                        self.__cv.wait()
                    if len(self.__q) == 0 and self.__working_count == 0:
                        return
                    from_url = self.__q.popleft()
                    if from_url is None:
                        break
                    self.__working_count += 1
                name = hostname(from_url)
                for to_url in htmlParser.getUrls(from_url):
                    if name != hostname(to_url):
                        continue
                    with self.__cv:
                        if to_url not in self.__lookup:
                           self.__lookup.add(to_url)
                           self.__q.append(to_url)
                           self.__cv.notify()
                with self.__cv:
                    self.__working_count -= 1
                    if not self.__q and not self.__working_count:
                        self.__cv.notifyAll()

        workers = []
        self.__q = collections.deque([startUrl])
        lookup = set([startUrl])
        for i in range(self.NUM_Threads):
            t = threading.Thread(target=worker, args=(htmlParser,))
            t.start()
            workers.append(t)

        for t in workers:
            t.join()
        return list(self.__lookup)    










import threading
from queue import Queue

class Solution:
    def crawl(self, startUrl: str, htmlParser: 'HtmlParser') -> List[str]:
        def hostname(url):
            start = len("http://")
            i = start
            while i < len(url) and url[i] != "/":
                i += 1
            return url[start:i]
        
        queue = Queue()
        seen = {startUrl}
        start_hostname = hostname(startUrl)
        seen_lock = threading.Lock()
        
        def worker():
            while True:
                url = queue.get()
                if url is None:
                    return

                for next_url in htmlParser.getUrls(url):
                    if next_url not in seen and hostname(next_url) == start_hostname:
                        seen_lock.acquire()
                        # Acquire lock to ensure urls are no enqueed multiple times
                        if next_url not in seen:
                            seen.add(next_url)
                            queue.put(next_url)
                        seen_lock.release()
                queue.task_done()
        
        num_workers = 8
        workers = []
        queue.put(startUrl)
        
        for i in range(num_workers):
            t = threading.Thread(target=worker)
            t.start()
            workers.append(t)
        
        # Wait until empty
        queue.join()
        
        for i in range(num_workers):
            queue.put(None)
        for t in workers:
            t.join()
        
        return list(seen)

File: Concurrency/1242. Web Crawler Multithreaded.cpp


/*

比如: 
["http://news.yahoo.com","http://news.yahoo.com/news","http://news.yahoo.com/news/topics/","http://news.google.com","http://news.yahoo.com/us"]
[[2,0],[2,1],[3,2],[3,1],[0,4]]
"http://news.yahoo.com/news/topics/"

比如有4个thread

Step 1:  下面用index 表示url

thread 0:   condition 因为 q.size() > 0, 继续  get 2 (q.top()) （release lock 让 2,3,4 开始）->  call getUrls -> push 2的friend 1 & 0  => notify_all
thread 1:  wait， 因为 q.size = 0
thread 2:  wait  因为 q.size = 0
thread 3:  wait  因为 q.size = 0


step2: 没有等待 thread 1 ,2 ,3 醒过来，thread 0开始下一轮工作,

thread 0: condition 因为 q.size() > 0, 继续  ;get 1 (q.top())  ->  call urls  (release lock 让 thread 1 可以开始)   
thread 1: wake up ->  发现 q size > 0,  get 2 (q.top()) -> call urls( release lock, 让3,4 可以开始)
thread 2: wake up ->  q.size = 0 -> sleep
thread 3: wake up ->  q.size = 0 -> sleep
thread 0: -> (check 3,4 过后) lock again ->  push 1的friend 4  => notify_all

step3: 没有等待 thread 1 ,2 ,3 醒过来，thread 0开始下一轮工作,

thread 0:  condition 因为 q.size() > 0, 继续  ; get 4 (q.top())  ->  call urls  (release lock 让 thread 1,2,3 可以开始) -> 
               
thread 2: wake up (来自step 2的notify_all) ->  q.size = 0  -> sleep 
thread 3: wake up (来自step 2的notify_all) ->  q.size = 0 -> sleep
thread 1: 从step 2中 lock again ->  2没有friend => notify all, 还有等到下一轮 thread 1工作, thread 0 先获取lock
thread 0:  (check 3,4 过后) 且 thread 1 完成后 lock again -> 4没有friend => notifyall 并更新done = true


step4: 没有等待 thread 1 ,2 ,3 醒过来，thread 0开始下一轮工作,

thread 0:   done = true, return  
thread 2:  wake up->   done = true, return  
thread 3: wake up->    done = true, return  
thread 1: wake up->     done = true, return  

注上面每一步thread 0, 1, 2,3 激活顺序不固定，

*/



/**
 * // This is the HtmlParser's API interface.
 * // You should not implement it, or speculate about its implementation
 * class HtmlParser {
 *   public:
 *     vector<string> getUrls(string url);
 * };
 */
class Solution {
private: 
    string hostname;
    int thread_num;
    queue<string>q; 
    unordered_set<string>seen;
    int working= 0;
    bool done;
    mutex m;
    condition_variable cv;
    
    string extract(const string& url){
        int start = url.find('/') + 2;
        int end = url.find('/',start);
        if(end == string::npos)
            return url.substr(start);
        return url.substr(start, end-start);
    }
     vector<thread>workers;
    
public:
    
    void worker(HtmlParser& parser, int threadNum){
        while(true){
            unique_lock<mutex>ul(m);
            cv.wait(ul, [&](){
               return q.size() > 0 || done;  
            });//被notify called后, 如果条件为true, 继续下面的，否则继续sleep
            if(done)
                return;
            working++;
            string cur = q.front(); q.pop();
            ul.unlock();
            
            auto nxt = parser.getUrls(cur);
            ul.lock();
            for(auto & url: nxt){
                if(seen.count(url) || extract(url)!=hostname)
                    continue;
                seen.insert(url);
                q.push(url);
            }
            working--;
            if(working == 0 && q.empty())
                done = true;
            cv.notify_all();
            
        }
        
        
    }
    
    vector<string> crawl(string startUrl, HtmlParser htmlParser) {
        hostname = extract(startUrl);
        done = false;
        thread_num = thread::hardware_concurrency();
        q.push(startUrl);
        seen.insert(startUrl);
        for(int i = 0; i < thread_num; i++){
            workers.emplace_back(&Solution::worker, this, ref(htmlParser), i); //每个thread 就开始工作了
        }
        
        // join those threads so that crawl is a blocking call 
        for(auto &t : workers){
            t.join();
        }
        // return every unique processed string
        return vector<string>(seen.begin(), seen.end());

    }
   
};

/**
 * // This is the HtmlParser's API interface.
 * // You should not implement it, or speculate about its implementation
 * class HtmlParser {
 *   public:
 *     vector<string> getUrls(string url);
 * };
 */
class Solution {

    mutex mu;
    condition_variable cv;
    queue<string>q;
    unordered_set<string>lookup;
    string host;
    int work_count = 0;
    vector<thread>threads;
    
    string extract(const string& url){
        int start = url.find('/') + 2;
        int end = url.find('/',start);
        if(end == string::npos)
            return url.substr(start);
        return url.substr(start, end-start);
    }
    
    void work(HtmlParser& htmlParser){
        while(true){
            string curUrl;
            {
                unique_lock<mutex>lock(mu);
                cv.wait(lock, [&](){return !q.empty() || work_count == 0;});
                if(q.empty() && work_count == 0)
                    break;
                ++work_count;
                curUrl = q.front();
                q.pop();
            }
            for(auto nxt: htmlParser.getUrls(curUrl)){
                if(extract(nxt)!=host) continue;
                {
                    unique_lock<mutex>lock(mu);
                    if(lookup.count(nxt) == 0){
                        q.push(nxt);
                        lookup.insert(nxt);
                        cv.notify_one(); //notify_one 就可以了 更efficient 只呼叫一个等待的thread
                    }
                }
            }
            {
                unique_lock<mutex>lock(mu);
                --work_count;
                if(q.empty() && work_count == 0)
                    cv.notify_all();
            }
        }
    }
    
    
public:
    vector<string> crawl(string startUrl, HtmlParser htmlParser) {
        host = extract(startUrl);
        q.push(startUrl);
        lookup.insert(startUrl);
        
        int numThreads = thread::hardware_concurrency();
        for(int i = 0; i<numThreads; ++i){
            threads.emplace_back(&Solution::work, this, ref(htmlParser));
        }
        for(int i = 0; i<numThreads; ++i)
            threads[i].join();
        return vector<string>(lookup.begin(),lookup.end());
    }
};


// Time:  O(|V| + |E|)
// Space: O(|V|)

/*

比如: 
["http://news.yahoo.com","http://news.yahoo.com/news","http://news.yahoo.com/news/topics/","http://news.google.com","http://news.yahoo.com/us"]
[[2,0],[2,1],[3,2],[3,1],[0,4]]
"http://news.yahoo.com/news/topics/"

比如有4个thread

Step 1:  下面用index 表示url

thread 0:  lock 1 acquire ->  get 2 (q.top()) -> ++ work_count  release lock get URL  
thread 1:  wait(relase lock)， 因为 q.size = 0
thread 2:  wait(relase lock)  因为 q.size = 0
thread 3:  wait(relase lock)  因为 q.size = 0
main:     lock 0 wait-> 因为 work_count > 0
thread 0: lock 2 acquire -> push 0 (2的friend)  => notifyall 
thread 0: 没等 其他thread wake up  lock 2 acquire -> push 1 (2的friend)  => notifyall 


step2: 

thread 3: wake up(Step 1 thread 0 的 notify_all) ->  lock 1 acquire ->  get 0 (q.top()) -> ++ work_count  release lock 1 get URL 
thread 1: wake up(Step 1 thread 0的 notify_all) ->  lock 1 acquire ->  get 1 (q.top()) -> ++ work_count   release lock 1 get URL
thread 0: lock 3 acquire -> --work_count ->  release lock 3 
          lock 1 ->  sleep 因为 qsize = 0  wait (relase lock1)
thread 2: wake up(Step 1的 notify_all)  因为 q.size = 0 -> 继续wait 
main:     lock 0 wait-> 因为 work_count > 0 


step 3：

thread 3: lock 2 acquire -> push 4 (0的friend)  => notifyall 
thread 0: wake up(step 3 thread 3的 notify_all) ->  lock 1 acquire -> get 4 (q.top()) ->  ++ work_count release lock get URL 
thread 3: lock 3 acquire -> --work_count ->  release lock 3 
          lock 1 ->  sleep 因为 qsize = 0  wait (relase lock1)
thread 2:  wait (step 3 thread 3的 notify_all)  因为 q.size = 0 -> 继续wait
thread 1: lock 2 acquire -> 1没有friend 不push 
          lock 3 acquire -> --work_count ->  release lock 3 
          lock 1 ->  sleep 因为 qsize = 0  wait (relase lock1)  


Step 4: 
thread 0: lock 2 acquire -> 1没有friend 不push 
          lock 3 acquire -> --work_count ->  release lock 3 
          因为 work_count == 0 且 workcount == 0: => notify_all
thread 1; wait (step 4 thread 0的 notify_all)  因为 q.size = 0 -> 继续wait
main:  waitup, lock 0 acquire -> push 四个空的 string 进 q

thread 2: wakeup -> cur = 空 return 
thread 1: wakeup -> cur = 空 return 
thread 3: wakeup -> cur = 空 return 
thread 0: wakeup -> cur = 空 return 



注上面每一步thread 0, 1, 2,3 激活顺序不固定，

*/


class Solution {
public:
    vector<string> crawl(string startUrl, HtmlParser htmlParser) {
        q_.emplace(startUrl);
        unordered_set<string> lookup = {startUrl};
        vector<thread> workers;
        for (int i = 0; i < NUMBER_OF_WORKERS; ++i) {
            workers.emplace_back(bind(&Solution::worker, this, &htmlParser, &lookup));
        }
        {
            unique_lock<mutex> lock{m_};  //lock 0
            cv_.wait(lock, [this]() { return q_.empty() && !working_count_; });
            for (const auto& t : workers) {
                q_.emplace();
            }
            cv_.notify_all();
        }
        for (auto& t : workers) {
            t.join();
        }
        return vector<string>(lookup.cbegin(), lookup.cend());
    }

private:
    void worker(HtmlParser *htmlParser, unordered_set<string> *lookup) {
        while (true) {
            string from_url;
            {
                unique_lock<mutex> lock{m_};  //lock 1
                cv_.wait(lock, [this]() { return !q_.empty(); });
                from_url = q_.front(); q_.pop();
                if (from_url.empty()) {
                    break;
                }
                ++working_count_;
            }
            const auto& name = hostname(from_url);
            for (const auto& to_url: htmlParser->getUrls(from_url)) {
                if (name != hostname(to_url)) {    
            // The function hostname itself is safe to call. It becomes dangerous if they're all trying to operate on the same variable.
                    continue;
                }
                unique_lock<mutex> lock{m_}; //lock 2
                if (!lookup->count(to_url)) {
                    lookup->emplace(to_url);
                    q_.emplace(to_url);
                    cv_.notify_all();
                }
            }
            {
                unique_lock<mutex> lock{m_}; // //lock 3
                --working_count_;
                if (q_.empty() && !working_count_) {
                    cv_.notify_all();
                }
            }
        }
    }
    
    string hostname(const string& url) {
        static const string scheme = "http://";
        return url.substr(0, url.find('/', scheme.length()));
    }

    static const int NUMBER_OF_WORKERS = 4;
    queue<string> q_;
    int working_count_ = 0;
    mutex m_;
    condition_variable cv_;
};




// Time:  O(|V| + |E|)
// Space: O(|V|)

/**
 * // This is the HtmlParser's API interface.
 * // You should not implement it, or speculate about its implementation
 * class HtmlParser {
 *   public:
 *     vector<string> getUrls(string url);
 * };
 */
class Solution {
public:
    vector<string> crawl(string startUrl, HtmlParser htmlParser) {
        q_.emplace(startUrl);
        unordered_set<string> lookup = {startUrl};
        vector<thread> workers;
        const auto& worker = [this](HtmlParser *htmlParser, unordered_set<string> *lookup) {
            while (true) {
                string from_url;
                {
                    unique_lock<mutex> lock{m_};
                    cv_.wait(lock, [this]() { return !q_.empty(); });
                    from_url = q_.front(); q_.pop();
                    if (from_url.empty()) {
                        break;
                    }
                    ++working_count_;
                }
                const auto& name = hostname(from_url);
                for (const auto& to_url: htmlParser->getUrls(from_url)) {
                    if (name != hostname(to_url)) {
                        continue;
                    }
                    unique_lock<mutex> lock{m_};
                    if (!lookup->count(to_url)) {
                        lookup->emplace(to_url);
                        q_.emplace(to_url);
                        cv_.notify_all();
                    }
                }
                {
                    unique_lock<mutex> lock{m_};
                    --working_count_;
                    if (q_.empty() && !working_count_) {
                        cv_.notify_all();
                    }
                }
            }
        };
        for (int i = 0; i < NUMBER_OF_WORKERS; ++i) {
            workers.emplace_back(worker, &htmlParser, &lookup);
        }
        {
            unique_lock<mutex> lock{m_};
            cv_.wait(lock, [this]() { return q_.empty() && !working_count_; });
            for (const auto& t : workers) {
                q_.emplace();
            }
            cv_.notify_all();
        }
        for (auto& t : workers) {
            t.join();
        }
        return vector<string>(lookup.cbegin(), lookup.cend());
    }

private:
    string hostname(const string& url) {
        static const string scheme = "http://";
        return url.substr(0, url.find('/', scheme.length()));
    }

    static const int NUMBER_OF_WORKERS = 4;
    queue<string> q_;
    int working_count_ = 0;
    mutex m_;
    condition_variable cv_;
};
